### PR TITLE
Defining HAL target backend interfaces as part of #1168.

### PR DESCRIPTION
**Pull request by benvanik**
This adds the new TargetBackend interface as sketched out in #1168. With this a bulk of the target-specific code for workgroup size/count and dispatch logic moves into this interface and support is extended for multiple target binaries. Serialization is now performed separately which makes the target backend code much easier to read and keeps the binary blobs out of the IR until the very end.

There's still pending work here that will have to wait until upstream changes land for dynamic pass registration; right now the passes for each backend are run in a nested pass manager like the previous implementation however the design is setup such that we can yank out the iree-hal-translate-executables and iree-hal-serialize-executables passes when they are supported.

Future changes will clean up the hal.executable.binary op to use a more free-form identifier and match that up with the dispatch recording. For now we only have one executable per target backend so things are fine, but as soon as we start generating variants for different SPIR-V versions/etc we'll have to adjust that.

### CODE REVIEW

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMTarget.cpp [hunk 1]
```diff
@@ -120,118 +123,112 @@ struct PrepareForLLVMLoweringPass
...
+                                    OpBuilder& executableBuilder) override {
+    // LLVM is not thread safe and currently translation shares an LLVMContext.
+    // Since we serialize executables from multiple threads we have to take a
+    // global lock here.
+    static llvm::sys::SmartMutex<true> mutex;
+    llvm::sys::SmartScopedLock<true> lock(mutex);
+
+    // At this moment we are leaving MLIR LLVM dialect land translating module
+    // into target independent LLVMIR.
+    // targetOp.getInnerModule().dump();
```
Comment (asaadaldien) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMTarget.cpp:179]: delete this comment. leftover debugging ?

**Review by asaadaldien — COMMENTED** [review]
Very cool stuff! high level question what linkExecutable is going to do ?

**Comment by benvanik** [comment]
> Very cool stuff! high level question what linkExecutable is going to do ?

I wanted to chat with you about that :)
The intent is to allow target backends to perform arbitrary combination of any number of translated executables. They'll basically get the list of executables they generated and be able to return sets that should be linked together. The scheduling code will then rewrite all the dispatches to point at those new grouped executables.

This means if you have @executable_dispatch_a with an entry point a() and @executable_dispatch_b with entry point b() you could link those together into @executable_dispatch_ab with entry points a() and b(). If you think then of each executable in the LLVM case being an object file (one entry point function and maybe some other functions) then the linkage step lets you perform linking between those (deduplication and other LTO). For the AOT case this means you can now produce one or more shared objects/archives instead of one per executable.

SPIR-V could use this too to reduce code size, as if a lot of duplication is present the executables could be combined and functions reused. Even if the driver isn't smart about handling the multiple entry points during compilation it would reduce the fixed overhead of SPIR-V files and reduce runtime work (small number of interactions with the driver API and cache instead of one per executable).

There's still a lot of details to work out, but there's now a nice place in the pipeline to put things and we can start thinking about it :)

File: iree/compiler/Dialect/HAL/Transforms/TranslateExecutables.cpp [hunk 2]
```diff
@@ -14,105 +14,69 @@
...
-        }
-
-        // Perform translation using the registered translation function.
-        if (failed(targetFn->second(executableOp, executableOptions_))) {
-          executableOp.emitError()
-              << "failed translation to target " << targetBackend;
+    auto executableOp = getOperation();
+    auto targetOps = llvm::to_vector<4>(
+        executableOp.getBlock().getOps<IREE::HAL::ExecutableTargetOp>());
+    for (auto targetOp : targetOps) {
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Transforms/TranslateExecutables.cpp:45]: That's the goal, however today the pass system cannot use any information about the ops to determine whether to run. So nesting on ExecutableTargetOp for the SPIR-V linalg path would also run the nested xla path would also run vmla/etc.

Mehdi and River have a Q2 OKR to enable dynamic pass execution so that we can nest on ExecutableTargetOp *and* a function that checks attributes (like whether this target executable is for SPIR-V, whether it's using the linalg path, etc). At that point I'll be able to remove this pass entirely and have the buildTransformationPipeline in the TargetBackend just add that dynamic pass registration itself. Same with serialization and linking.

File: iree/compiler/Dialect/HAL/Target/TargetBackend.h [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+      IREE::HAL::ExecutableTargetOp targetOp, OpBuilder &executableBuilder) {
+    llvm_unreachable("unimplemented serializeExecutable");
+    return failure();
+  }
+
+ protected:
+  // Calculates the workgroup size (x, y, z).
+  virtual std::array<Value, 3> calculateDispatchWorkgroupSize(
+      Location loc, IREE::HAL::ExecutableOp executableOp,
+      IREE::HAL::ExecutableEntryPointOp entryPointOp, Value workload,
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.h:287]: The problem is that if we fix a single shaped workload that implies that we have a single shaped output and we want dispatch regions to be able to have many outputs. That's how things were for awhile but I wanted to remove that limitation: the workload now just denotes how much work is to be done but not the shape of the work. The shape should be based on the actual shape of the I/O and operations, and all of those are available here (entryPointOp.signature has the original tensor<> signature). That gives the backends more control over how they interpret shapes and compute workgroup counts (whether they need to be in tiles or in items, how bounds checking/padding happens, etc).

For dynamic shapes we also don't know the exact shapes (or maybe anything but the rank) when first translating the executable and even if we could wait until we saw the dispatch that's not enough: thanks to dynamic shapes every single dispatch could differ so there's no data we should be relying on for correctness there (just potential specialization info).

The nice thing is that only the default recordDispatch implementation calls these functions; if you are doing your own recordDispatch then you can get access to the exact shapes of all I/O (even if dynamic), though by the point you are recording dispatches you've likely already fixed your workgroup size.

I left some TODOs in here with notes of where you may want to change this in SPIR-V-land: https://github.com/google/iree/blob/d1db8b4721cc50c70eb5be2007e7c2a1a1d181f7/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L167-L179

This will be useful when we want to support runtime-chosen workgroup sizes as specialization constants - but for now this is at least an improvement over one workgroup size for all backends and all dispatches :)

TL;DR: a backend can do whatever it wants now but there are no assumptions made in IREE about the shape of work or how that work is tiled. Linalg-on-tensors as a tiling mechanism running prior to IREE may be able to do much smarter things and then just pass around the workload for IREE to schedule things, but the final linalg-to-blah lowering could still access the info from the original linalg-on-tensor pass that includes shapes.

[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.h:287]: Also, what this frees you up for is looking at the actual ops inside the dispatch region to compute the value: is it a depthwise conv and what are the best values for that, etc - even if there's a reshape at the end fused in that changes the shape and would otherwise mess up logic. Some kind of cost modeling that is based on the context of the work being performed will let you do a lot of good things (or, learn them with magic ML 🦄  ;)

**Comment by MaheshRavishankar** [comment]
Change looks good to me overall. Touches a lot of build files and VMLA stuff which I dont know enough to review meaningfully. Sets up things nicely for the backends.

Thanks!

File: iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertVariableOps.cpp [hunk 4]
```diff
@@ -79,8 +79,13 @@ class VariableLoadOpConversion
   LogicalResult matchAndRewrite(
       IREE::HAL::VariableLoadOp op, llvm::ArrayRef<Value> operands,
       ConversionPatternRewriter &rewriter) const override {
-    rewriter.replaceOpWithNewOp<IREE::VM::GlobalLoadRefOp>(
-        op, typeConverter.convertType(op.getType()), op.variable());
+    if (IREE::VM::RefType::isCompatible(op.getType())) {
+      rewriter.replaceOpWithNewOp<IREE::VM::GlobalLoadRefOp>(
+          op, typeConverter.convertType(op.getType()), op.variable());
+    } else {
+      rewriter.replaceOpWithNewOp<IREE::VM::GlobalLoadI32Op>(
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertVariableOps.cpp:86]: today the only type besides ref types is i32, so this is all there is :)
